Source/WTF/wtf/CompactVariant.h

```diff
@@ -0,0 +1,163 @@
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/CompactVariantOperations.h>
+
+namespace WTF {
+
+// A `CompactVariant` acts like a `std::variant` with the following differences:
+// - All alternatives must be pointers, smart pointers, have size of 48 bits or fewer, or be specialized for `CompactVariantTraits`.
+// - It can only contain 254 or fewer alternatives.
+// - Has a more limited API, only offering `holds_alternative()` for type checking and `switchOn()` for value access.
+
+template<CompactVariantCapable V> class CompactVariant {
+    static_assert(std::variant_size_v<V> < 255);
+    using Variant = V;
+    using Index = uint8_t;
+    using Operations = CompactVariantOperations<V>;
+public:
+    template<typename U> constexpr CompactVariant(U&& value)
+        requires std::constructible_from<V, U> && (!std::same_as<std::remove_cvref_t<U>, CompactVariant>)
+    {
+        m_data = Operations::template encode<typename VariantBestMatch<V, U>::type>(std::forward<U>(value));
+    }
+
+    template<typename T, typename... Args> explicit constexpr CompactVariant(std::in_place_type_t<T>, Args&&... args)
+        requires std::constructible_from<V, T>
+    {
+        m_data = Operations::template encodeFromArguments<T>(std::forward<Args>(args)...);
+    }
+
+    template<size_t I, typename... Args> explicit constexpr CompactVariant(std::in_place_index_t<I>, Args&&... args)
+        requires std::constructible_from<V, std::variant_alternative_t<I, V>>
+    {
+        m_data = Operations::template encodeFromArguments<std::variant_alternative_t<I, V>>(std::forward<Args>(args)...);
+    }
+
+    CompactVariant(const CompactVariant& other)
+    {
+        Operations::copy(m_data, other.m_data);
+    }
+
+    CompactVariant& operator=(const CompactVariant& other)
+    {
+        if (*this == other)
+            return *this;
+
+        Operations::destruct(m_data);
+        Operations::copy(m_data, other.m_data);
+
+        return *this;
+    }
+
+    CompactVariant(CompactVariant&& other)
+    {
+        Operations::move(m_data, other.m_data);
+
+        // Set `other` to the "moved from" state.
+        other.m_data = UINT64_MAX;
```

Comment:
Would be nicer to use C++ instead of C style `UINT64_MAX` and maybe also a named constant for this special value.